libyul/backends/evm/StackHelpers.h

```diff
@@ -377,6 +377,147 @@ class Shuffler
 	}
 };
 
+/// An simple optimized map for mapping StackSlots to ints.
+class Multiplicity
+{
+public:
+	Multiplicity(size_t _initialCapacity):
+		m_functionCallReturnLabelSlotMultiplicity(_initialCapacity),
+		m_functionReturnLabelSlotMultiplicity(0),
+		m_variableSlotMultiplicity(_initialCapacity),
+		m_junkSlotMultiplicity(0) {}
+
+	int& operator[](StackSlot const& _slot)
+	{
+		if (auto* p = std::get_if<FunctionCallReturnLabelSlot>(&_slot))
+			return m_functionCallReturnLabelSlotMultiplicity[*p];
+		if (std::holds_alternative<FunctionReturnLabelSlot>(_slot))
+			return m_functionReturnLabelSlotMultiplicity;
+		if (auto* p = std::get_if<VariableSlot>(&_slot))
+			return m_variableSlotMultiplicity[*p];
+		if (auto* p = std::get_if<LiteralSlot>(&_slot))
+			return m_literalSlotMultiplicity[*p];
+		if (auto* p = std::get_if<TemporarySlot>(&_slot))
+			return m_temporarySlotMultiplicity[*p];
+		return m_junkSlotMultiplicity;
+	}
+
+	int at(StackSlot const& _slot) const
+	{
+		if (auto* p = std::get_if<FunctionCallReturnLabelSlot>(&_slot))
+			return m_functionCallReturnLabelSlotMultiplicity.at(*p);
+		if (std::holds_alternative<FunctionReturnLabelSlot>(_slot))
+			return m_functionReturnLabelSlotMultiplicity;
+		if (auto* p = std::get_if<VariableSlot>(&_slot))
+			return m_variableSlotMultiplicity.at(*p);
+		if (auto* p = std::get_if<LiteralSlot>(&_slot))
+			return m_literalSlotMultiplicity.at(*p);
+		if (auto* p = std::get_if<TemporarySlot>(&_slot))
+			return m_temporarySlotMultiplicity.at(*p);
+		return m_junkSlotMultiplicity;
+	}
+
+private:
+	template <class T>
+	class StackSlotHashMap
+	{
+	public:
+		StackSlotHashMap(size_t _initialCapacity)
+		{
+			m_capacity = 0;
+			m_size = 0;
+			m_initialCapacity = 2;
+			while (m_initialCapacity < _initialCapacity)
+				m_initialCapacity *= 2;
```

Comment:
Shouldn't the allocation (`resize`) happen here?

Reply:
Often, nothing will be stored in the hashmap. We don’t want to allocate and initialize memory in such cases.